Faster Call for `CompiledSDFG`

philip-paul-mueller opened the pull request:
This PR adds a new possibility how a `CompiledSDFG` can be called.

Before `__call__()` performed a lot of operations, which can be summarized as:
- Ensuring that the arguments are ordered correctly.
- Transforming them to the right type (`ndarray` to pointers).

Especially when benchmarking smaller SDFG this is a lot of overhead, which actually dominates the execution.
Furthermore, the runtime heavily depends on the number of arguments.

To solve this, this PR introduces the `_fast_call()` function, which expects that its arguments are already in the right order and casted to the right type.
In addition it does some refactoring and splits `_construct_args()` into multiple parts.

philip-paul-mueller commented:
@BenWeber42 Now ready for review.

BenWeber42 commented:
Hi, thanks for the PR!

Unfortunately, I'm not too familiar with these parts of the codebase, so it would take me quite some time to do a proper review. Let's see if we can find someone who could do the review more efficiently...

tbennun commented:
@philip-paul-mueller general comment before the review: please remove the unrelated (NaN) changes from the PR.

philip-paul-mueller commented:
@tbennun Sorry that thing got somehow mixed up.

tbennun reviewed (changes requested):
The idea for having an argument-check-free and/or pregenerated argument fastcall is good. However, the PR may introduce different performance issues. The original code was written to be "fast" (using tuples, comprehensions, and fewer function calls in the original code), and the new fastcall should be just as fast.

Additionally, it sounds like this should be an external API for use by users (otherwise the PR does not make much sense), but the call is prefixed with an underscore. Moreover, some of the code (preparing outputs) should also not be in the context of the fastcall. See comments.

philip-paul-mueller commented on dace/codegen/compiled_sdfg.py:
```diff
@@ -438,33 +473,93 @@ def _construct_args(self, kwargs) -> Tuple[Tuple[Any], Tuple[Any]]:
+
+        try:
+            # Replace arrays with their base host/device pointers
+            newargs = []
```
Before I added `_fast_call()` I tried to improve `_construct_args()` I profiled it and the function as a whole was faster (but still too slow).
However, since I care more on `_fast_call()` I could revert these changes.

tbennun replied:
Please do.

philip-paul-mueller replied:
See https://github.com/spcl/dace/pull/1467#discussion_r1436812693 for reason why not.

philip-paul-mueller commented on dace/codegen/compiled_sdfg.py:
```diff
@@ -366,21 +366,56 @@ def _get_error_text(self, result: Union[str, int]) -> str:
-                    self._cfunc(self._libhandle, *argtuple)
+                    self._cfunc(self._libhandle, *callargs)
 
             if self.has_gpu_code:
```
I do not understand this. According to my understand this calls the actual compiled function. Thus without it nothing would be done, or do I miss something here?

tbennun replied:
Look at the line the comment points to: the GPU runtime check belongs in the normal call, not fast call. If you want to ensure execution is fast you can skip this check, which might be expensive.

philip-paul-mueller replied:
Thanks for the explanation, so the comment always points to the "last line that is shown"?

philip-paul-mueller commented on dace/codegen/compiled_sdfg.py:
```diff
@@ -438,33 +473,93 @@ def _construct_args(self, kwargs) -> Tuple[Tuple[Any], Tuple[Any]]:
         for i, (a, arg, atype) in enumerate(zip(argnames, arglist, argtypes)):
-            if not dtypes.is_array(arg) and isinstance(atype, dt.Array):
+            arg_is_array = dtypes.is_array(arg)
+            atype_is_dtArray = isinstance(atype, dt.Array)
```
During my profiling I saw that these functions cost a lot of time, so I cached the results.
And what part of the guideline you mean, the naming?

tbennun replied:
Naming of the variable.

philip-paul-mueller commented on dace/codegen/compiled_sdfg.py:
```diff
@@ -366,21 +366,56 @@ def _get_error_text(self, result: Union[str, int]) -> str:
+            )
+        argtuple, initargtuple = self._construct_args(kwargs)   # Missing arguments will be detected here.
+                                                                #  Return values cached in `self._lastargs`.
+        return self._fast_call(argtuple, initargtuple)
```
I am not sure what you mean by that. since `_fast_call()` returns the result of `_convert_return_values()` `__call__()` will return the same. So I do not see a need to call it again.

philip-paul-mueller commented on dace/codegen/compiled_sdfg.py:
```diff
@@ -366,21 +366,56 @@ def _get_error_text(self, result: Union[str, int]) -> str:
+            self,
+            callargs: Tuple[Any, ...],
+            initargs: Tuple[Any, ...],
+    ) -> Any:
```
I reword the docstring, but I am not sure if you will like it.

philip-paul-mueller commented:
Thanks for your work and especially for correcting my typos.
I have addressed most of your suggestions, but for some I am unsure (see comment).

Regarding your comments about my splitting of `_construct_args()`, I would be fine to revert this if you would prefer that.
When I did the split (and all further modifications) I constantly checked its runtime and according to that it is now faster.
One reason is, that it spend less time with type checking since the results are cached.

@tbennun

tbennun commented on dace/codegen/compiled_sdfg.py:
```diff
@@ -503,39 +576,39 @@ def _construct_args(self, kwargs) -> Tuple[Tuple[Any], Tuple[Any]]:
-                           if not symbolic.issymbolic(arg) or (hasattr(arg, 'name') and arg.name not in constants))
+    def _construct_args_callparams(self, arglist, arg_ctypes, argtypes, argnames):
+        constants  = self.sdfg.constants
+        callparams = []
```
List creation and dynamically expanding its size can be expensive, performance wise.

tbennun commented on dace/codegen/compiled_sdfg.py:
```diff
@@ -407,23 +453,40 @@ def __del__(self):
             kwargs[desc[0]] = arr
 
-        # Argument construction
+        arglist, argtypes, argnames = self._construct_args_arglist(kwargs)
```
Function calls can be expensive, performance-wise.
If you think they are not, please show some runtime results that say otherwise.

philip-paul-mueller replied:
As requested I performed some experiment.
I used the attached function, which is based on the `spmv` tutorial (randomly selected, if I should test another function please provide one).

I got the following results for the current master (`09d37e9e`):
```
min: 0.0001241610007127747s
max: 0.0007202319975476712s
mean: 0.0001336343956439426s
std: 3.586189058597028e-05
q_{1, 5, 25, 50, 75, 95, 99}: [0.00012455 0.00012489 0.00012555 0.00012643 0.00012851 0.00014832 0.00037179]
```

For my previous state, denoted append version, (`eb0198a5`) I got the following results:
```
min: 0.00010327700147172436s
max: 0.0006421329999284353s
mean: 0.00011245609261338056s
std: 3.2908162087660314e-05
q_{1, 5, 25, 50, 75, 95, 99}: [0.00010416 0.00010479 0.0001056 0.00010633 0.00010738 0.00012325 0.00030589]
```

And for the current version (`ff3ced35`):
```
min: 0.00010886099698836915s
max: 0.0006989000030444004s
mean: 0.00011409736964318047s
std: 2.089365416625181e-05
q_{1, 5, 25, 50, 75, 95, 99}: [0.00010925 0.00010962 0.00011023 0.000111 0.00011243 0.00011744 0.00020909]
```

As you can see from these results even the append version is faster than the current master.
According to my memory the majority of these gains come from caching some type checking results.
We also see that the new version is a little bit slower (in case of `min`) than the append version, but the differences are very small, but its 99 percentile is much lower.
I think this should suffice to show that this version is not slower than the original version.
[call_args.py.gz](https://github.com/spcl/dace/files/13776544/call_args.py.gz)

philip-paul-mueller replied:
Because you insisted so much on it I decided to put everything back into one function (`1a1fca4b`).
However, I also run the test again with the following results:
```
min: 0.00010630900214891881s
max: 0.0006737619987688959s
mean: 0.00011147427259614536s
std: 2.1202032309282628e-05
q_{1, 5, 25, 50, 75, 95, 99}: [0.00010679 0.00010708 0.00010762 0.00010819 0.00010998 0.00011565 0.00020852]
```
As you can see there is not much difference compared to the previous (`ff3ced35`) version where the `_construct_args()` function was separated into smaller pieces.

tbennun reviewed (commented):
Thank you for addressing the formatting changes. However, my major comments were still not addressed.

philip-paul-mueller commented:
Hello Tal,
thanks for your patience with me, I have addressed all your comments.
Except one I have removed all `.append()` loops, the one I kept had the advantages of fusing two loops.
I also did the benchmarking you requested, [see below](https://github.com/spcl/dace/pull/1467#discussion_r1436812693), which show that the new code is at least as fast as the original one and reverting would hurt performance.

I which you a happy new year.

Best, Philip

tbennun reviewed (approved):
LGTM
